service/history/workflow/update/update.go

```diff
@@ -27,84 +27,358 @@ package update
 )
 
 const (
-	statePending state = iota
+	stateAdmitted state = iota
+	stateProvisionallyRequested
+	stateRequested
+	stateProvisionallyAccepted
 	stateAccepted
-	stateRejected
+	stateProvisionallyCompleted
 	stateCompleted
 )
 
-func (s state) String() string {
-	switch s {
-	case statePending:
-		return "pending"
-	case stateAccepted:
-		return "accepted"
-	case stateRejected:
-		return "rejected"
-	case stateCompleted:
-		return "completed"
-	default:
-		return "unknown"
+// New creates a new Update instance with the provided ID that will call the
+// onComplete callback when it completes.
+func New(id string, onComplete func(), opts ...updateOpt) *Update {
+	upd := &Update{
+		id:              id,
+		state:           stateAdmitted,
+		onComplete:      onComplete,
+		instrumentation: &noopInstrumentation,
+		accepted:        future.NewFuture[*failurepb.Failure](),
+		outcome:         future.NewFuture[*updatepb.Outcome](),
+	}
+	for _, opt := range opts {
+		opt(upd)
 	}
+	return upd
 }
 
-func newUpdate(request *updatepb.Request, protocolInstanceID string) *Update {
-	return &Update{
-		state:              statePending,
-		request:            request,
-		messageID:          uuid.New(),
-		protocolInstanceID: protocolInstanceID,
-		outcome:            future.NewFuture[*updatepb.Outcome](),
+func withInstrumentation(i *instrumentation) updateOpt {
+	return func(u *Update) {
+		u.instrumentation = i
 	}
 }
 
+func newAccepted(id string, onComplete func(), opts ...updateOpt) *Update {
+	upd := &Update{
+		id:              id,
+		state:           stateAccepted,
+		onComplete:      onComplete,
+		instrumentation: &noopInstrumentation,
+		accepted:        future.NewReadyFuture[*failurepb.Failure](nil, nil),
+		outcome:         future.NewFuture[*updatepb.Outcome](),
+	}
+	for _, opt := range opts {
+		opt(upd)
+	}
+	return upd
+}
+
+func newCompleted(
+	id string,
+	fetchOutcome func(ctx context.Context) (*updatepb.Outcome, error),
+	opts ...updateOpt,
+) *Update {
+	upd := &Update{
+		id:              id,
+		state:           stateCompleted,
+		instrumentation: &noopInstrumentation,
+		accepted:        future.NewReadyFuture[*failurepb.Failure](nil, nil),
+		outcome:         lazyOutcome(fetchOutcome),
+	}
+	for _, opt := range opts {
+		opt(upd)
+	}
+	return upd
+}
+
+// WaitOutcome observes this Update's completion, returning the Outcome when it
+// is available. This call will block until the Outcome is known or the provided
+// context.Context expires.
 func (u *Update) WaitOutcome(ctx context.Context) (*updatepb.Outcome, error) {
 	return u.outcome.Get(ctx)
 }
 
-func (u *Update) accept() {
-	u.state = stateAccepted
+// WaitOutcome observes this Update's acceptance, returning a Failure if the
+// update is rejected or nil if it is successfully accepted. This call will
+// block until the Outcome is known or the provided context.Context expires.
+func (u *Update) WaitAccepted(ctx context.Context) (*failurepb.Failure, error) {
```

Comment:
I'm thinking now that this should return `(*updatepb.Outcome, error)`. Reasoning is that (a) we may be able to optimize so that if this update has been accepted and _also_ completed then we can return a surprise outcome here, and (b) the caller is just going to turn the Failure into an Outcome anyway so we might as well do it here. I know (b) because I'm just now implementing support for async invocation in updateworkflow.Invoke.